# DNA Sequence Manipulation and Analysis

## Usage
Program executable is 'main.out'. main.out takes in exactly one argument, which is the file that contains the sequence of the amino acid to be analyzed:
```
main.out [filename]
```
The file's contents must have a header on the first line, and the sequence to be analyzed on the second line. File test1.txt is provided as an example. Unfortunately, program is only coded to allow for sequences in all uppercase letters. 

## Important Notes
* Need to have "AminoAcids.txt" in same folder as executable. The file contains the information that will be read in and stored into a Linear Probing Hash Table. The entries are references that will be used to create AminoAcids.

* Structural form of an amino acid sequence is limited to about 18 amino acids, this approximation is based off of the width of the programmer's laptop screen. If a sequence that contains a gene that codes for more than ~18 amino acids, the structure won't be displayed properly due to the implementation of the structure.

* Information/data of invalid genes (sequences that don't contain a start codon and/or a stop codon) won't be printed to the screen. They will show up as empty lines. The only exception to this would be nucleotide sequences, which are printed regardless. 

* The addition of proline to the chain won't create a "kink" like it's supposed to in the planar structure generated by the program. A "non-kink" structure of the proline is used instead to stay consistent with the other amino acids and make the code for the output of the planar structure easier.

* If user decides to use the testprint() method in structure to print out a structure of their choice, be aware that if the user chooses to pick Tryptophan as the first amino acid, part of its structure will be cut off in the grid. Due to its R-group containing two rings, its structure takes up more space width wise than all the other amino acids. (Normally not a problem since "viable" genes start with Methionine)

* All functions are declared and organized in alphabetical order to allow the programmer to find a specific function quicker. Brief descriptions of each function (unless trivial/obvious) is given in the comments that precede the function in its respective .cpp file.

## Future Improvements
* Add throws/exceptions
* Allow for chains longer than 18 amino acids by adding extra "boxes" and allowing the chain to wrap down
* Adjust so that it doesn't print the "grid" for the structure when it's not a viable gene
* Allow for uppercase and/or lowercase letters in the DNA sequence given in the file

## Authors
Alissa Tran, Nancy Nguyen
